### ISSUE TITLE
Add author attribution in info.json

### ISSUE THREAD

**Issue by wadefagen** [PrairieLearn/PrairieLearn#1824]
Currently, the `info.json` file in each PLv3 question does not provide any first-class support for authors, editors, or other forms of attributions.  Based on conversations with several colleagues, nearly all "mature" courses provide some form of attribution (author tags, external metadata, or even comments in the Python/HTML files) -- but this is inconsistent across courses and has no best practices due to the ad-hoc nature.

The addition of attributions for content creators would allow for professional recognition for faculty, TAs, CAs, and other content creators, would provide for better accountability for poorly created questions, and provide a way to reach out to the correct person when questions arise about a specific question.  It would also seem natural for the author(s) to be available to content consumers (eg: students).  A fantastic report I would love to have is to find out how many times a TA I'm writing a recommendation letter for has had their questions generated for a student -- for some of my top TAs, I imagine this would be tens of thousands (!!)

In my use cases, two different levels would be minimally necessary:

- `author(s)` (string | array), attribution for the primary developer(s) of the question content
- `editor(s)` (string | array), attribution for fixing typos, minor bugs, or upgrading questions from previous versions

**Comment by echuber2** [comment]
Seems reasonable to me. I know some content gets reused for years and misattributed eventually. (At least, I know of one undergrad CA who had developed a complex MP from scratch, only to have a subsequent semester TA replace the name in the credits.)

**Comment by trombonekenny** [comment]
Yeah, I was going to say that authorship and attribution is best mined from the git history. That's where I would start to study the origins and modifications of questions.

**Comment by echuber2** [comment]
A script could auto-populate based on that, but there are probably cases where content was moved from its original development repo manually and the git history might not retain that forever depending on what happens, then. Git blame works for staff but there's a possibility the one who formatted it for PL pulled it from course materials that weren't originally made for PL, too, i.e. the actual author isn't automatically revealed that way. This proposal would also formalize some way to have attribution in the student-facing UI itself, if that's appropriate.

**Comment by mwest1066** [comment]
I really like this idea! Having a "user" page like GitHub that shows statistics (number of questions generated, etc) feels critical to allow proper professional credit.

I'm currently thinking about a format like:
```
"contributors": [
    {"uid": "waf@illinois.edu", "date": "2018-07-12", "action": "authored"},
    {"uid": "mwest@illinois.edu", "date": "2018-07-14", "action": "tested"},
    {"uid": "zilles@illinois.edu", "date": "2019-01-30", "action": "edited", comment: "converted to v3"}
]
```

I feel like it's good to give lots of opportunities for credit? Linux kernel patches have `Signed-off-by `, `Acked-by`, `Co-Developed-by`, `Reported-by`, `Tested-by`, `Reviewed-by`, `Suggested-by`, and arbitrary user-created tags.

An initial list of possible actions for PL questions is:

Action | Description
--- | ---
`authored` | Created the actual question text or the generation/grading code. Did not just re-implement a question that someone else wrote. It's fine to have multiple authors for a question if more than one person substantially contributed.
`implemented` | Took a question from a different source (e.g., a paper question or another LMS) and implemented it in PL.
`edited` | Fixed typos or minor bugs, or upgraded questions from a previous version.
`reviewed` | Gave feedback on the question wording, implementation, or pedagogy. Did not test the question for correctness (use `tested` for this). The review feedback might have simply been "no changes needed, question is good".
`tested` | Tested the question for bugs and answer correctness. Did not have to find any problems or make any changes, and may have just concluded that the question is ready to go.
`reported` | Reported a valid problem with a question in some context other than testing. For example, a student reporting a valid bug on a question during an exam.

Is the date information useful, or is that just going to be a pain? I'm anticipating that we'd add an in-browser interface to add these "contribution" fields, so the date could be added automatically.

Are there any good ideas for interfacing this information with the git history? I'll admit that my initial thought on reading this Issue was "it's all in git", but I now see that this is not true (e.g., the git history does not distinguish between authoring and editing). My current thinking is that this information is just completely separate from the git history.

**Comment by mfsilva22** [comment]
I think date is useful. And I think this a a great idea, and very much needed. I just hope we can find a way of encouraging people to use it.

**Comment by mwest1066** [comment]
@wadefagen following up on your comment about authors being available to students, are you thinking that the _student_ question page would display "Question author: Wade Fagen"? In talking about this with a few people, it seems like some instructors don't want this visible, so I guess we'll need to have a property like `"display": True` on the author data.

Should this default to displaying or not displaying author information on student question pages? 

I'm assuming we will always show author information to instructors and we'll always include all data in summary statistics (e.g., "how many times has this question been generated") so we don't need flags to control this.

**Comment by SethPoulsen** [comment]
I'm going to close this, as the relevant discussion was superseded by this: https://github.com/PrairieLearn/PrairieLearn/issues/10004, which was later superseded by this: https://github.com/PrairieLearn/PrairieLearn/issues/13603
